File: python/14_merge_intervals.py
#!/usr/bin/env python3
"""
LeetCode #56: 合并区间
题目: 给定若干区间，合并所有重叠的区间。
解法: 按左端点排序，然后依次合并 - 维护当前区间 [cur_l, cur_r]，
      若下一个区间的左端点 <= cur_r，则合并（cur_r = max(cur_r, next_r)）；
      否则保存当前区间，开始新区间。时间 O(n log n)，空间 O(1)。
"""


def merge(intervals: list[list[int]]) -> list[list[int]]:
    """
    排序 + 合并: 按左端点排序，遍历合并重叠区间。
    """
    if not intervals:
        return []
    intervals.sort(key=lambda x: x[0])
    result = [[intervals[0][0], intervals[0][1]]]
    for l, r in intervals[1:]:
        if l <= result[-1][1]:          # 有重叠，合并
            result[-1][1] = max(result[-1][1], r)
        else:                           # 无重叠，开新区间
            result.append([l, r])
    return result


if __name__ == "__main__":
    print(merge([[1, 3], [2, 6], [8, 10], [15, 18]]))  # [[1,6],[8,10],[15,18]]
    print(merge([[1, 4], [4, 5]]))                    # [[1,5]]

File: cpp/14_merge_intervals.cpp
/**
 * LeetCode #56: 合并区间
 * 题目: 合并所有重叠的区间。
 * 解法: 按左端点排序，然后依次合并 - 维护当前区间，
 *       若下一区间左端点 <= cur_r 则合并，否则开新区间。
 * 时间 O(n log n)，空间 O(1)。
 */

#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        if (intervals.empty()) return {};
        sort(intervals.begin(), intervals.end(),
             [](const vector<int>& a, const vector<int>& b) {
                 return a[0] < b[0];
             });
        vector<vector<int>> result;
        result.push_back(intervals[0]);
        for (size_t i = 1; i < intervals.size(); ++i) {
            if (intervals[i][0] <= result.back()[1])
                result.back()[1] = max(result.back()[1], intervals[i][1]);
            else
                result.push_back(intervals[i]);
        }
        return result;
    }
};

int main() {
    Solution sol;
    vector<vector<int>> intervals = {{1,3},{2,6},{8,10},{15,18}};
    auto res = sol.merge(intervals);
    for (auto& v : res)
        cout << "[" << v[0] << "," << v[1] << "] ";
    cout << endl;
    return 0;
}
